### ISSUE TITLE
aot_elf_release takes a unreasonably long time to complete

### ISSUE THREAD

**Issue by TrueHorse** [flutter/flutter#150182]
### Steps to reproduce

1. Create a new Flutter project
2. Add hyphenatorx as a dependency (`flutter pub add hyphenatorx`)
3. Add `Language_de_1996();` as the first line in the main.dart build method
4. Run `flutter build windows -v`

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:hyphenatorx/languages/language_de_1996.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    Language_de_1996();
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // TRY THIS: Try running your application with "flutter run". You'll see
        // the application has a purple toolbar. Then, without quitting the app,
        // try changing the seedColor in the colorScheme below to Colors.green
        // and then invoke "hot reload" (save your changes or press the "hot
        // reload" button in a Flutter-supported IDE, or press "r" if you used
        // the command line to start the app).
        //
        // Notice that the counter didn't reset back to zero; the application
        // state is not lost during the reload. To reset the state, use hot
        // restart instead.
        //
        // This works for code too, not just values: Most code changes can be
        // tested with just a hot reload.
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // TRY THIS: Try changing the color here to a specific color (to
        // Colors.amber, perhaps?) and trigger a hot reload to see the AppBar
        // change color while the other colors stay the same.
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          //
          // TRY THIS: Invoke "debug painting" (choose the "Toggle Debug Paint"
          // action in the IDE, or press "p" in the console), to see the
          // wireframe for each widget.
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headlineMedium,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
```

</details>

### What target platforms are you seeing this bug on?

Windows

### OS/Browser name and version | Device information

Edition	Windows 10 Pro
Version	22H2
Installiert am	‎28.‎05.‎2021
Betriebssystembuild	19045.4529
Leistung	Windows Feature Experience Pack 1000.19058.1000.0

Prozessor	Intel(R) Core(TM) i3-4170 CPU @ 3.70GHz   3.70 GHz
Installierter RAM	16,0 GB (15,9 GB verwendbar)
Geräte-ID	56DA7894-60B4-47E7-944F-917DD0AC0E39
Produkt-ID	00330-71344-19537-AAOEM
Systemtyp	64-Bit-Betriebssystem, x64-basierter Prozessor
Stift- und Toucheingabe	Für diese Anzeige ist keine Stift- oder Toucheingabe verfügbar.

### Does the problem occur on emulator/simulator as well as on physical devices?

Unknown

### Logs

Log from sample app above:
[sample_build.txt](https://github.com/user-attachments/files/15818327/sample_build.txt)

Log from my app:
[completed build.txt](https://github.com/user-attachments/files/15816889/completed.build.txt)

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, 3.22.0, on Microsoft Windows [Version 10.0.19045.4412], locale de-DE)
[√] Windows Version (Installed version of Windows is version 10 or higher)
[√] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
[√] Chrome - develop for the web
[√] Visual Studio - develop Windows apps (Visual Studio Community 2022 17.8.6)
[√] Android Studio (version 2022.3)
[√] VS Code, 64-bit edition (version 1.74.3)
[√] Connected device (3 available)
[√] Network resources

• No issues found!

```

</details>

The Problem occures in line 779 in the log of the sample app and in line 271 in the other log, where aot_elf_release takes about one hour to complete. The error at the end was my bad.
This only happens using certain language classes of hyphenatorx. I tested all german languange codes, Hungarian, Finnish, Dutch and US-English and only Finnish and English finished in a reasonable amount of time.
Building in debug mode works normally.

[Issue in Hyphenatorx](https://github.com/xErik/hyphenatorx/issues/6)

**Comment by danagbemava-nc** [comment]
Reproducible using the code sample provided above. This does not appear to be specific to the windows platform as I try to build on macOS and iOS and on both platforms, the build got stuck at different points in the process.

This might be an issue with the dart compiler itself.

Labeling for further investigation. 

 
[ios.log](https://github.com/user-attachments/files/15820692/ios.log)
[macos.log](https://github.com/user-attachments/files/15820693/macos.log)

[build_ios_master.log](https://github.com/user-attachments/files/15820691/build_ios_master.log)

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 3.22.2, on macOS 14.4.1 23E224 darwin-arm64, locale en-GB)
    • Flutter version 3.22.2 on channel stable at /Users/nexus/dev/sdks/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 761747bfc5 (8 days ago), 2024-06-05 22:15:13 +0200
    • Engine revision edd8546116
    • Dart version 3.4.3
    • DevTools version 2.34.3

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /Users/nexus/Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.3)
    • Xcode at /Applications/Xcode-15.3.0.app/Contents/Developer
    • Build 15E204a
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.1)
    • Android Studio at /Users/nexus/Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.2.5)
    • IntelliJ at /Users/nexus/Applications/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 77.2.2
    • Dart plugin version 232.10286

[✓] VS Code (version 1.89.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.90.0

[✓] Connected device (5 available)
    • Nexus (mobile)                  • 00008020-001875E83A38002E • ios            • iOS 17.5.1 21F90
    • Dean’s iPad (mobile)            • 00008103-000825C811E3401E • ios            • iOS 17.5.1 21F90
    • macOS (desktop)                 • macos                     • darwin-arm64   • macOS 14.4.1 23E224 darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad     • darwin         • macOS 14.4.1 23E224 darwin-arm64
    • Chrome (web)                    • chrome                    • web-javascript • Google Chrome 125.0.6422.142

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

```
[!] Flutter (Channel master, 3.23.0-13.0.pre.212, on macOS 14.4.1 23E224 darwin-arm64, locale en-GB)
    • Flutter version 3.23.0-13.0.pre.212 on channel master at /Users/nexus/dev/sdks/flutters
    ! Warning: `flutter` on your path resolves to /Users/nexus/dev/sdks/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutters. Consider adding /Users/nexus/dev/sdks/flutters/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/nexus/dev/sdks/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutters. Consider adding /Users/nexus/dev/sdks/flutters/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 7b6d667701 (13 hours ago), 2024-06-12 19:34:52 -0400
    • Engine revision c7fcbfce60
    • Dart version 3.5.0 (build 3.5.0-250.0.dev)
    • DevTools version 2.36.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /Users/nexus/Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.3)
    • Xcode at /Applications/Xcode-15.3.0.app/Contents/Developer
    • Build 15E204a
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.1)
    • Android Studio at /Users/nexus/Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.2.5)
    • IntelliJ at /Users/nexus/Applications/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 77.2.2
    • Dart plugin version 232.10286

[✓] VS Code (version 1.89.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.90.0

[✓] Connected device (5 available)
    • Nexus (mobile)                  • 00008020-001875E83A38002E • ios            • iOS 17.5.1 21F90
    • Dean’s iPad (mobile)            • 00008103-000825C811E3401E • ios            • iOS 17.5.1 21F90
    • macOS (desktop)                 • macos                     • darwin-arm64   • macOS 14.4.1 23E224 darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad     • darwin         • macOS 14.4.1 23E224 darwin-arm64
    • Chrome (web)                    • chrome                    • web-javascript • Google Chrome 125.0.6422.142

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```
</details>

**Comment by mraleph** [comment]
There might be something we can change in the compiler - but I would recommend changing representation of language data instead. The current way things are represented is not very reasonable - having multimegabyte map literal in the source file --- among other things it will produce ungodly amount of native code.

Simplest change you could make is to convert it to string and JSON parse it, even better would be to use some more optimized representation.

**Comment by jpohhhh** [comment]
@mraleph TL;DR: optimized representation meaning something like protos? some bytestream format that optimizes parsing for whatever qualities i desire? (ex. runtime ram vs. cpu vs. compressed size)

context:

- Today I've been chasing down *only* release builds "hanging", timing out after 3x their normal length, on all native platforms

- I'm flummoxed, diffed a `flutter build --verbose` log from last passing build on CI and there's nothing different

- I came here to browse through issues, and I saw your comment, and realized this roughly coincides with me introducing a 200K line map literal, and I'm thinking now, at the very least, I should clean that up. 

- examples below, in case they're helpful as an absurdist edge case. trivia: They're used to match a list of ints (bytes) to an int (token), implementing tokenization for LLMs

IntListKey.dart: https://gist.github.com/jpohhhh/9878269d015e31530d70484e23ba5f9f
200K: https://gist.github.com/jpohhhh/f12aa40bcab87275696f5edcfaec4a54
100K: https://gist.github.com/jpohhhh/cfb6f64de905ce437081eb96709dc953

**Comment by jpohhhh** [comment]
The Maps did end up being the issue creating long / never finishing builds.
Removing them caused build times to fall to 1/3rd of their previous value (ex. 28 m => 8m iOS, 34 minutes => 11m Android)
For various reasons, loading a file/asset was going to cause issues (TL;DR Squadron x loading files off the main isolate x widget bindings to have access to root bundle)
Simply moving from a Map produced from a file, to a String with the file contents parsed at runtime, retained the working build + build time improvements. There is no visible impact at runtime.

**Comment by brianquinlan** [comment]
I think that this is related to the issue: https://github.com/xErik/hyphenatorx/issues/6

**Comment by mraleph** [comment]
This will be fixed by https://dart-review.googlesource.com/c/sdk/+/373502

**Comment by mraleph** [comment]
The problem reported by @TrueHorse is now fixed upstream by dart-lang/sdk@eb88654

The problem reported by @jpohhhh is different, I have made a separate patch for that: https://dart-review.googlesource.com/c/sdk/+/373661

**Comment by mraleph** [comment]
Both reported issues should now be fixed upstream.
